src/interrupt/kvm_irq/pci_msi_irq.rs

```diff
@@ -0,0 +1,223 @@
+// Copyright (C) 2019 Alibaba Cloud. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Manage virtual device's PCI MSI/PCI MSIx interrupts based on Linux KVM framework.
+//!
+//! To optimize for performance by avoiding unnecessary locking and state checking, we assume that
+//! the caller will take the responsibility to maintain the interrupt states and only issue valid
+//! requests to this driver. If the caller doesn't obey the contract, only the current virtual
+//! machine will be affected, it shouldn't break the host or other virtual machines.
+
+use super::msi_irq::{create_msi_routing_entries, new_msi_routing_entry, MsiConfig};
+use super::*;
+use std::os::unix::io::AsRawFd;
+
+pub(super) struct PciMsiIrq {
+    base: InterruptIndex,
+    count: InterruptIndex,
+    vmfd: Arc<VmFd>,
+    irq_routing: Arc<KvmIrqRouting>,
+    msi_configs: Vec<MsiConfig>,
+}
+
```

Comment:
Does each device have a kind of such group like `PciMsiIrq` including in the device specific structure?

Reply:
Something like this:
```
    fn update_msi_cfg(&mut self) -> Result<()> {
        self.intr_mgr
            .set_msi_low_address(self.msi_index_select, self.msi_address_low)?;
        self.intr_mgr
            .set_msi_high_address(self.msi_index_select, self.msi_address_high)?;
        self.intr_mgr
            .set_msi_data(self.msi_index_select, self.msi_data)?;
        if self.intr_mgr.is_enabled() {
            self.intr_mgr.modify(self.msi_index_select)?;
        }
        Ok(())
    }
```

Reply:
Thanks for the demo codes explaining it. I am considering if `self` means a device backend for example, it will hold the entire irq routing table information, right? Is it a good way?

Reply:
Yes, self is a backend device.